cmd/juju/model/defaultscommand.go

```diff
@@ -97,120 +131,425 @@ func (c *defaultsCommand) SetFlags(f *gnuflag.FlagSet) {
+			return false, errors.Trace(err)
+		}
+	} else {
+		cTag = names.NewCloudTag(cloudName)
+	}
+	cloud, err = cc.Cloud(cTag)
+	if err != nil {
+		return false, errors.Trace(err)
+	}
+
+	for _, r := range cloud.Regions {
+		if r.Name == region {
+			isCloudRegion = true
+			break
+		}
+	}
+	return isCloudRegion, nil
+}
+
+// parseSetKeys iterates over the args and make sure that the key=value pairs
+// are valid. It also checks that the same key isn't also being reset.
+func (c *defaultsCommand) parseSetKeys(args []string) error {
+	options, err := keyvalues.Parse(args, true)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	keyCounter := map[string]int{}
+	c.values = make(attributes)
+	for k, v := range options {
+		if k == config.AgentVersionKey {
+			return errors.Errorf(`%q must be set via "upgrade-juju"`, config.AgentVersionKey)
+		}
+		keyCounter[k]++
+		c.values[k] = v
+	}
+	for _, resetKey := range c.resetKeys {
+		keyCounter[resetKey]++
+		if keyCounter[resetKey] > 1 {
+			return errors.Errorf(`key %q specified more than once`, resetKey)
+		}
 	}
-	c.action = c.getDefaults
 	return nil
+
 }
 
-// getAPI sets the api on the command. This allows passing in a test
-// ModelDefaultsAPI implementation.
-func (c *defaultsCommand) getAPI() (defaultsCommandAPI, error) {
-	if c.api != nil {
-		return c.api, nil
+// handleSetArgs parses args for setting defaults.
+func (c *defaultsCommand) handleSetArgs(args []string) error {
+	argZeroKeyOnly := !strings.Contains(args[0], "=")
+	// If an invalid region was specified, the first positional arg won't have
+	// an "=". If we see one here we know it is invalid.
+	switch {
+	case argZeroKeyOnly && c.regionName == "":
+		return errors.Errorf("invalid region specified: %q", args[0])
+	case argZeroKeyOnly && c.regionName != "":
+		return errors.New("cannot set and retrieve default values simultaneously")
+	default:
+		if err := c.parseSetKeys(args); err != nil {
+			return errors.Trace(err)
+		}
+		c.action = c.setDefaults
+		return nil
 	}
+}
 
-	api, err := c.NewAPIRoot()
-	if err != nil {
-		return nil, errors.Annotate(err, "opening API connection")
+// handleZeroArgs determines whether we are doing a reset only and if any
+// additional arguments are valid -- or why they are not.
+func (c *defaultsCommand) handleZeroArgs() error {
+	resetSpecified := c.resetKeys != nil
+	regionSpecified := c.regionName != ""
+
+	switch {
+	case regionSpecified && !resetSpecified:
+		// It doesn't make sense to specify a region unless we are resetting
+		// values here.
+		return errors.New("specifying a region when retrieving defaults is invalid")
+	case regionSpecified && resetSpecified:
+		// We can reset for a region, c.action stays nil. We want do reset
+		// only.
+		return nil
+	case !regionSpecified && resetSpecified:
+		// Again we want to reset, this is for resetting the controller as
+		// there is no region specified. So c.action remains nil.
+		return nil
+		// case !regionSpecified && !resetSpecified:
```

Comment:
I meant to leave a note in for future reference, because if we didn't short circuit we would hit that case here. Seems useful to have a note explaining why it isn't here. I clarified without leaving the code block in wholesale.